rust/src/livefs.rs

```diff
@@ -0,0 +1,381 @@
+const LIVEFS_STATE_NAME: &str = "rpmostree-livefs-state";
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+struct LiveFsState {
+    /// The OSTree commit that the running root filesystem is using,
+    /// as distinct from the one it was booted with.
+    commit: Option<String>,
+    /// Set when a livefs operation is in progress; if the process
+    /// is interrupted, some files from this commit may exist
+    /// on disk but in an incomplete state.
+    inprogress_commit: Option<String>,
+}
+
+fn get_runstate_dir(deploy: &ostree::Deployment) -> PathBuf {
+    format!(
+        "{}/{}.{}",
+        OSTREE_RUNSTATE_DIR,
+        deploy.get_csum().expect("csum"),
+        deploy.get_deployserial()
+    )
+    .into()
+}
+
+// fn get_transient_dir(deploy: &ostree::Deployment) -> Result<PathBuf> {
+//     let key = match deploy.get_unlocked() {
+//         DeploymentUnlockedState::Transient => "unlocked-transient",
+//         DeploymentUnlockedState::Development => "unlocked-development",
+//         s => bail!("Unhandled unlock state {}", s),
+//     };
+//     let p = get_runstate_dir(deploy);
+//     Ok(std::fs::read_to_string(p.join(key))?.into())
+// }
+
+fn get_livefs_state(deploy: &ostree::Deployment) -> Result<Option<LiveFsState>> {
+    let root = openat::Dir::open("/")?;
+    if let Some(f) = root.open_file_optional(&get_runstate_dir(deploy).join(LIVEFS_STATE_NAME))? {
+        let s: LiveFsState = serde_json::from_reader(std::io::BufReader::new(f))?;
+        Ok(Some(s))
+    } else {
+        Ok(None)
+    }
+}
+
+fn write_livefs_state(deploy: &ostree::Deployment, state: &LiveFsState) -> Result<()> {
+    let rundir = get_runstate_dir(deploy);
+    let rundir = openat::Dir::open(&rundir)?;
+    rundir.write_file_with(LIVEFS_STATE_NAME, 0o644, |w| -> Result<_> {
+        Ok(serde_json::to_writer(w, state)?)
+    })?;
+    Ok(())
+}
+
+// TODO use this later
+// fn get_transient_repo(booted: &ostree::Deployment) -> Result<ostree::Repo> {
+//     use std::os::unix::io::AsRawFd;
+//     let tdir = get_transient_dir(booted)?;
+//     let tdir = openat::Dir::open(&tdir)?;
+//     let opts = glib::VariantDict::new(None).to_variant();
+//     let repo = ostree::Repo::create_at(tdir.as_raw_fd(), "repo", ostree::RepoMode::Bare, &opts, gio::NONE_CANCELLABLE)?;
+//     Ok(repo)
+// }
+
+fn relpath_dir(p: &Path) -> Result<&Path> {
+    Ok(p.strip_prefix("/")?.parent().expect("parent"))
+}
+
+fn subpath(diff: &crate::ostree_diff::FileTreeDiff, p: &Path) -> Option<PathBuf> {
+    if let Some(ref d) = diff.subdir {
+        let p = p.strip_prefix("/").expect("prefix");
+        Some(Path::new(d).join(p))
+    } else {
+        Some(p.to_path_buf())
+    }
+}
+
+fn apply_diff(
+    repo: &ostree::Repo,
+    diff: &crate::ostree_diff::FileTreeDiff,
+    commit: &str,
+    destdir: &openat::Dir,
+) -> Result<()> {
+    if !diff.changed_dirs.is_empty() {
+        anyhow::bail!("Changed directories are not supported yet");
+    }
```

Comment:
Well it's not hard, it just requires a direct implementation (or perhaps we could add something like `ostree_update_directory_metadata_at()` or something?)

Basically it can't be done in terms of ostree checkouts.